### ISSUE TITLE
pytorch port or pretrained model conversion

### ISSUE THREAD

**Issue by adrienchaton** [pranaymanocha/PerceptualAudio#6]
Dear Pranay,

I am working on generative models for musical sound synthesis. As other related works, the optimization of the generation uses spectrogram reconstructions (e.g. multiscale spectral loss of DDSP or NSF).

The models perform rather well but there usually are still some "just noticeable" artifacts in the synthesized sounds. And I came across your paper, I think it is really interesting to try your differentiable metric for musical sound synthesis, which suffer from such fine artifacts.

I am working on pytorch, and not familiar with tensorflow unfortunately ..
Are there any plans to port your codes or convert your pretrained models to pytorch ?

I would try otherwise .. Maybe some help is needed, and I'd like to keep in touch about that if you are interested. Converting the pretrained scratch_loss might be simpler than retraining with all codes in pytorch.

**Comment by pranaymanocha** [comment]
Hey!
Thanks for the message. Unfortunately, I am currently tied up in other projects which is why I cannot find the time to write the pytorch version of the repo. But I will be extremely happy if someone in the community is interested and will be more than happy to clarify in case of problems. 

Thanks!

**Comment by adrienchaton** [comment]
Great, thank you !
I am working in the next days on coding it in Pytorch, you already provide the data in a well structured manner, and some details in the paper, that's very helpful. I will ask you some more informations when needed and put everything on github. And keep you informed when it's done.

**Comment by adrienchaton** [comment]
Hi Pranay,

I started and you can have a look here if time allows:
https://github.com/adrienchaton/PerceptualAudio_pytorch

Regarding the small classifier, I just added a low-capacity MLP that infers the binary human label from the scalar distance D(x_ref,x_per). For the rest, I tried to stay ~ close to what your paper states. However, I am working at 22kHz as the models I develop are meant to synthesize audio at this sampling rate.

I am happy for any feedbacks you may give to this re-implementation.
I'll work on it more when I can, I just run some first trainings and do basic evaluations, which of course do not account for the whole analysis you do in your paper.

Thanks !

**Comment by pranaymanocha** [comment]
Hi Adrien,
Thanks for building it up. As for the comments, let me get back to you soon. One other thing that I observed regarding training is that as the simplest case, you can do gradient descent with batch size =1 (updating weights after seeing every example). As this adds noise to the training, it helps train the perceptual metric. 
As for your own task, I would recommend using my pretrained loss function and seeing if that works well for your artifacts.

**Comment by adrienchaton** [comment]
Hi Pranay,

Indeed, several papers have pointed the benefits of smaller batch sizes, improving generalization with some "regularization" from gradient noise and avoiding suboptimal solutions. However this trains slower. I updated and run trainings at batch size = 16 as in your code. I'll see like that and if needed, go below 16.

Regarding the possibility to port your pretrained loss to Pytorch, which I would love to. Unfortunately it kind of seems a bit complicate. I had a look at https://github.com/microsoft/MMdnn which in theory allows converting TF to Pytorch, but just trying to run the TensorFlow code already threw some errors, probably a version mismatch or some other details with my TF setup. Somehow, anytime I try to run some other's TF codes it gets complicated .. to be honest, and it is only my own little opinion, I like very much Pytorch API. So I found more straight-forward to try to retrain all the way in Pytorch (and also have a 22kHz model) .. If anyone around has experience with converting pretrained TF to Pytorch, please help :-)

Thanks for your feedbacks !

**Comment by adrienchaton** [comment]
Hi Pranay,

I am still experimenting on the model training, and when it seems satisfying I will try to add it up as a loss to the generative model I am working on. So far my only evaluation is the train/test set model accuracy in the binary prediction vs ground truth human judgement "similar" / "different".

How did your model performed on that ? What test set accuracy did you reach please ?

**Comment by pranaymanocha** [comment]
Hi Adrien,
I am extremely sorry: I have been slightly busy lately and finding it difficult to find time to review your codebase.
As for the loss, I was able to get test loss down to ~0.50 (as far as I can remember). 
As for your second question, the aim was to normalize the distances between [0,1] but later I found that the accuracy improves if I just remove that activation as it tries to squish all distances into a small space with little variance. I suggest you dont include the sigmoid activation on the distance before passing it on to the small classification network.  I had almost forgotten about this so thanks for clarifying this! :)

**Comment by adrienchaton** [comment]
Thank you Pranay for the feedback.
Do you remember roughly which binary accuracy did your test loss ~0.50 corresponded to ?

I am experimenting different variants, either with sigmoid, tanh (stronger gradient close to zero) or no activation. Let's see if I can get good results :-)
(so far it had tendency to overfit, even with small batch size)

I slightly updated codes, I hope you can have a look sometimes, any remark or hint are great !

**Comment by pranaymanocha** [comment]
Three comments:
1) I observed that initialization matters a lot. For training one model that works well on (the held out test set), I had to run approximately 5 runs with the same hyperparameters. This is because the space of human perception is large and we are trying to map it. There are a lot of local minima in this space and chances are that the model get struck and that depends on the initialization. Normalization matters a lot as well. I trained models with/without BatchNorm and found large differences. I suggest you try 4-5 different runs of the same hyperparameters.
2) As we are collecting data through AMT, training data is expected to be noisy (no matter how much precautions we may take). Introducing some noise to the model helps, but our data itself is noisy which is why adding too much noise (in terms of dropout) doesn't do very well.
3) I did not check binary accuracy. I feel that if you are able to get the test loss down to around 0.55, it should work well. ~0.69 CE is random performance. How much down are you able to currently bring down the test loss to?

Thanks,
Pranay

**Comment by adrienchaton** [comment]
Thanks for your reply Pranay,

I agree about the feel that initialization seems quite sensitive.
And also I stick to using batchnorm on both the distance (convolutional) network and the classification (dense) network.

The models start with ~0.69 CE loss which is consistent with what you confirm.
However so far I am a bit stuck with:
_ either models do not fit well and stay around 0.69CE both train/test set
_ either models overfit, train CE gets very low and accuracy close to 1 while test CE increases and accuracy hits a ~0.7 accuracy roof

An other observation from my side, depending on hyper-parameters (and init), it seems that sometimes the first backprop iteration does not give any gradient in the distance (convolutional) network layers. I always check the gradient magnitudes in network layers in the first backward, but here I don't identify yet what causes it and which parameters combinations may be avoided ...

**Comment by adrienchaton** [comment]
On the last point, actually I reconsider what I said about having stronger gradients close to 0 distance, e.g. using Tanh (to discriminate more subtle differences). I think the gradient init was going null because of the saturation of the Tanh or Sigmoid. Since the distance is evaluated on a high-dimensional feature space (all layer activations) there must be some chances of having strong "outliers", with large distance at init. Saturation would then ignore them in the optimization, causing more potential for being trapped in local minima and having poor generalization.
I am running some trainings with exponential shape activations, that would emphasize gradients for larger distances ..

**Comment by adrienchaton** [comment]
Hi Pranay,
Right now I get down to 0.5762 test loss and the model doesn't overfit anymore.
I don't run multiple times the same hyper-parameter setting to keep experimenting different variations, but best test loss results seem to work better with exponential shaped distance activation. It trains slower but seems to keep learning rather continuously. It could actually train longer and reach 0.55 or less maybe. I let you know how it goes ..

**Comment by adrienchaton** [comment]
Hi Pranay,

A little update, as I still experiment on using the perceptual audio metric.
I am also updating the codes on the github repo of the re-implementation.

So, on the side of training the metric, I still struggle with fitting well the test set and 0.5762 is the best I could get to. Either it tends to overfitting or it hardly learns.

Then I also started applying the best model I could get as an additional reconstruction cost for the generative model (a VQ-VAE on waveform) which already has latent regularization (discrete embedding) and multiscale spectrogram magnitude reconstruction cost.

I attach a plot of the average perceptual distance on the train/test set reconstructions of the generative model (dataset are clarinet performances) along epochs of optimization.

On the test set it is not optimizing, but on the train set it seems to behave well. What I find interesting is that in the first epochs it does not optimize since the generation is very wrong and is rather learning a rough fit with the spectrogram objective. And my guess is that around 200 epochs, the model starts reconstructing something decent which then also starts to refine with the perceptual metrics.

Perceptual evaluation, I mean my own listening of final test set reconstructions, tends to make me feel it's overall improving the fine detail artifacts, even though the learning curve on the test set doesn't optimize ... and it would be quite awesome to manage to fit well the loss model, what I still work on !

![percep_dist_plot](https://user-images.githubusercontent.com/35500385/76389203-179f5a00-63ae-11ea-864b-c1fc09ff0479.png)

**Comment by pranaymanocha** [comment]
Hi Adrien,
Thanks for the follow-up. I am glad that the perceptual model helps your model. I too have been facing these problems while training. I suspect that the extra noise using AMT makes the model a little unstable to train. 
As for the second comment, I can suggest two things:
1) How are you loading the files? As you may have seen, I have used wavfile.read (using scipy.io). If you look at the docs, you will see that if it finds the audio format in int16, it loads files using that range [-32768 to +32767]. So if you train the loss function by having audio in the range of [-1 to 1], you should use the same loading mechanism while using that loss model in the additional reconstruction cost for the generative model. Just make sure that you consistently use the same audio loading mechanism for training and eval - and try to maintain the ranges as well.

2) I tried using probability as a proxy for the loss. So if you visualize the classification layer, it was almost linear in the middle (slight exp in the beginning and ends). So I think unless the classification layer learns something that is high dimensional, we will have similar performance using probability as a proxy for loss.

Thanks,
Pranay

**Comment by adrienchaton** [comment]
Hi Pranay,

You are pointing something I didn't mind, thank you for that !
It explains the issue I guess, basically I convert all files from .wav with librosa.core.load and only apply the resampling/mono arguments. The data is saved in float32, then when I launch a training I just go select the desired .npy audio files.

And as part of my training hyper-parameters, I choose settings for silence trimming, amplitude normalization etc.

Now I inspected the datasets I combine (because well isolated instrument performances from multitrack datasets are scarce), and as expected they have quite different recording volumes. So if I don't apply amplitude normalization, ranges are very disparate, while the perceptual model is trained on more homogeneous volumes.

One drawback of using amplitude normalization across datasets is ... background noise amplification, and sometimes it's a bit severe.

Brings to my mind an other idea, I'll train the perceptual model with an additional random amplitude shifting in the training dataloader. As for time shift invariance, I'll apply a random gain to the pairs (same gain to both xref,xper). Let's see, maybe it fixes the subsequent training of the generative model without amplitude normalization and also prevents a bit more overfitting from the perceptual model training ..

Thanks ! :)

**Comment by pranaymanocha** [comment]
Hey Adrien,
Do you think I could

**Comment by pranaymanocha** [comment]
Hi Adrien,
I am closing this issue for now, but please feel free to reopen this if you have more questions/comments. It would also be great if you could upload your best model on your github page so that I could experiment with it on some data.

Thanks!
Pranay
